File: apis/sources/seismic.mjs
// USGS Earthquake Catalog — Nepal and the Main Himalayan Thrust
// Free, no key. Nepal sits on the collision boundary that produced the
// 2015 Gorkha earthquake, so this is the highest-consequence feed in the stack.

import { safeFetch, daysAgo } from '../utils/fetch.mjs';
import { SEISMIC_BBOX, PROVINCES, provinceOf, CITIES } from '../utils/nepal.mjs';

const BASE = 'https://earthquake.usgs.gov/fdsnws/event/1/query';

export async function getQuakes({ days = 30, minMagnitude = 2.5 } = {}) {
  const params = new URLSearchParams({
    format: 'geojson',
    starttime: daysAgo(days),
    minlatitude: String(SEISMIC_BBOX.lamin),
    maxlatitude: String(SEISMIC_BBOX.lamax),
    minlongitude: String(SEISMIC_BBOX.lomin),
    maxlongitude: String(SEISMIC_BBOX.lomax),
    minmagnitude: String(minMagnitude),
    orderby: 'time',
  });
  return safeFetch(`${BASE}?${params}`, { timeout: 20000 });
}

function haversineKm(lat1, lon1, lat2, lon2) {
  const R = 6371;
  const toRad = d => (d * Math.PI) / 180;
  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);
  const a = Math.sin(dLat / 2) ** 2
    + Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLon / 2) ** 2;
  return 2 * R * Math.asin(Math.sqrt(a));
}

function nearestCity(lat, lon) {
  let best = null;
  for (const c of Object.values(CITIES)) {
    const km = haversineKm(lat, lon, c.lat, c.lon);
    if (!best || km < best.km) best = { label: c.label, km: Math.round(km) };
  }
  return best;
}

function compactQuake(f) {
  const [lon, lat, depth] = f.geometry?.coordinates || [];
  const p = f.properties || {};
  return {
    id: f.id,
    mag: p.mag,
    place: p.place,
    time: p.time ? new Date(p.time).toISOString() : null,
    lat, lon,
    depthKm: depth,
    province: provinceOf(lat, lon),
    nearest: (typeof lat === 'number' && typeof lon === 'number') ? nearestCity(lat, lon) : null,
    felt: p.felt || 0,
    tsunami: Boolean(p.tsunami),
    url: p.url || null,
  };
}

export async function briefing() {
  const data = await getQuakes({ days: 30, minMagnitude: 2.5 });

  if (data?.error) {
    return { source: 'Seismic', timestamp: new Date().toISOString(), error: data.error };
  }

  const quakes = (data.features || []).map(compactQuake).filter(q => q.mag != null);
  const last24h = quakes.filter(q => q.time && Date.now() - new Date(q.time) < 86400000);
  const last7d = quakes.filter(q => q.time && Date.now() - new Date(q.time) < 7 * 86400000);

  const byProvince = {};
  for (const q of quakes) {
    if (!q.province) continue;
    byProvince[PROVINCES[q.province].label] = (byProvince[PROVINCES[q.province].label] || 0) + 1;
  }

  const strongest = [...quakes].sort((a, b) => b.mag - a.mag)[0] || null;
  const significant = quakes.filter(q => q.mag >= 4.5).slice(0, 15);

  // Shallow quakes do far more damage than deep ones at the same magnitude.
  const shallowStrong = quakes.filter(q => q.mag >= 4.0 && q.depthKm != null && q.depthKm < 35);

  const signals = [];
  if (strongest && strongest.mag >= 5.0) {
    signals.push(`M${strongest.mag} earthquake — ${strongest.place}`);
  }
  if (last24h.length >= 5) {
    signals.push(`${last24h.length} quakes in 24h — elevated seismic sequence`);
  }
  if (shallowStrong.length >= 3) {
    signals.push(`${shallowStrong.length} shallow M4+ events (<35km) — higher surface damage potential`);
  }
  if (last7d.some(q => q.mag >= 6.0)) {
    signals.push('M6+ event this week — expect aftershock sequence and infrastructure damage reports');
  }

  return {
    source: 'Seismic',
    timestamp: new Date().toISOString(),
    window: '30d',
    totalEvents: quakes.length,
    events24h: last24h.length,
    events7d: last7d.length,
    maxMagnitude: strongest?.mag ?? null,
    strongest,
    byProvince,
    significant,
    recent: quakes.slice(0, 25),
    signals,
  };
}

if (process.argv[1]?.endsWith('seismic.mjs')) {
  const data = await briefing();
  console.log(JSON.stringify(data, null, 2));
}

File: backend/app/domains/hazards/sources/seismic.py
"""USGS Earthquake Catalog — Nepal and the Main Himalayan Thrust.

Free, no key. Nepal sits on the collision boundary that produced the 2015
Gorkha earthquake, so this is the highest-consequence feed in the stack.

Runnable alone:  python -m app.domains.hazards.sources.seismic
"""

import math
from datetime import datetime, timezone
from typing import Any

from app.core.http import days_ago, is_error, now_iso, safe_fetch
from app.core.nepal import CITIES, PROVINCES, SEISMIC_BBOX, province_of

BASE = "https://earthquake.usgs.gov/fdsnws/event/1/query"

EARTH_RADIUS_KM = 6371


async def get_quakes(days: int = 30, min_magnitude: float = 2.5) -> Any:
    return await safe_fetch(
        BASE,
        timeout=20.0,
        params={
            "format": "geojson",
            "starttime": days_ago(days),
            "minlatitude": SEISMIC_BBOX.lamin,
            "maxlatitude": SEISMIC_BBOX.lamax,
            "minlongitude": SEISMIC_BBOX.lomin,
            "maxlongitude": SEISMIC_BBOX.lomax,
            "minmagnitude": min_magnitude,
            "orderby": "time",
        },
    )


def _haversine_km(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    d_lat = math.radians(lat2 - lat1)
    d_lon = math.radians(lon2 - lon1)
    a = (
        math.sin(d_lat / 2) ** 2
        + math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) * math.sin(d_lon / 2) ** 2
    )
    return 2 * EARTH_RADIUS_KM * math.asin(math.sqrt(a))


def _nearest_city(lat: float, lon: float) -> dict[str, Any] | None:
    best: dict[str, Any] | None = None
    for city in CITIES.values():
        km = _haversine_km(lat, lon, city.lat, city.lon)
        if best is None or km < best["km"]:
            best = {"label": city.label, "km": round(km)}
    return best


def _compact(feature: dict[str, Any]) -> dict[str, Any]:
    coords = (feature.get("geometry") or {}).get("coordinates") or []
    lon = coords[0] if len(coords) > 0 else None
    lat = coords[1] if len(coords) > 1 else None
    depth = coords[2] if len(coords) > 2 else None
    props = feature.get("properties") or {}

    time_ms = props.get("time")
    when = None
    if time_ms:
        moment = datetime.fromtimestamp(time_ms / 1000, tz=timezone.utc)
        when = f"{moment:%Y-%m-%dT%H:%M:%S}.{moment.microsecond // 1000:03d}Z"

    nearest = (
        _nearest_city(lat, lon)
        if isinstance(lat, (int, float)) and isinstance(lon, (int, float))
        else None
    )
    return {
        "id": feature.get("id"),
        "mag": props.get("mag"),
        "place": props.get("place"),
        "time": when,
        "lat": lat,
        "lon": lon,
        "depthKm": depth,
        "province": province_of(lat, lon),
        "nearest": nearest,
        "felt": props.get("felt") or 0,
        "tsunami": bool(props.get("tsunami")),
        "url": props.get("url"),
    }


def _age_ms(iso: str | None) -> float:
    if not iso:
        return float("inf")
    try:
        moment = datetime.fromisoformat(iso.replace("Z", "+00:00"))
    except ValueError:
        return float("inf")
    return (datetime.now(timezone.utc) - moment).total_seconds() * 1000


async def briefing() -> dict[str, Any]:
    data = await get_quakes(days=30, min_magnitude=2.5)

    if is_error(data):
        return {"source": "Seismic", "timestamp": now_iso(), "error": data.error}

    features = data.get("features") or [] if isinstance(data, dict) else []
    quakes = [q for q in (_compact(f) for f in features) if q["mag"] is not None]

    last_24h = [q for q in quakes if _age_ms(q["time"]) < 86_400_000]
    last_7d = [q for q in quakes if _age_ms(q["time"]) < 7 * 86_400_000]

    by_province: dict[str, int] = {}
    for q in quakes:
        if not q["province"]:
            continue
        label = PROVINCES[q["province"]].label
        by_province[label] = by_province.get(label, 0) + 1

    strongest = max(quakes, key=lambda q: q["mag"], default=None)
    significant = [q for q in quakes if q["mag"] >= 4.5][:15]

    # Shallow quakes do far more damage than deep ones at the same magnitude.
    shallow_strong = [
        q
        for q in quakes
        if q["mag"] >= 4.0 and q["depthKm"] is not None and q["depthKm"] < 35
    ]

    signals: list[str] = []
    if strongest and strongest["mag"] >= 5.0:
        signals.append(f"M{strongest['mag']} earthquake — {strongest['place']}")
    if len(last_24h) >= 5:
        signals.append(f"{len(last_24h)} quakes in 24h — elevated seismic sequence")
    if len(shallow_strong) >= 3:
        signals.append(
            f"{len(shallow_strong)} shallow M4+ events (<35km) — "
            "higher surface damage potential"
        )
    if any(q["mag"] >= 6.0 for q in last_7d):
        signals.append(
            "M6+ event this week — expect aftershock sequence and "
            "infrastructure damage reports"
        )

    return {
        "source": "Seismic",
        "timestamp": now_iso(),
        "window": "30d",
        "totalEvents": len(quakes),
        "events24h": len(last_24h),
        "events7d": len(last_7d),
        "maxMagnitude": strongest["mag"] if strongest else None,
        "strongest": strongest,
        "byProvince": by_province,
        "significant": significant,
        "recent": quakes[:25],
        "signals": signals,
    }


if __name__ == "__main__":
    import asyncio
    import json

    print(json.dumps(asyncio.run(briefing()), indent=2, ensure_ascii=False))
